packages/lib/src/acting-io.ts

```diff
@@ -0,0 +1,170 @@
+		if (itemIdOrIndex.length === 0) {
+			return false
+		}
+		const matchingItem = list.find(item => {
+			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+			// @ts-ignore
+			return (primaryKeyName in item) && itemIdOrIndex === item[primaryKeyName]
+		})
+		if (matchingItem) {
+			return itemIdOrIndex
+		}
+
+		if (!isIndexArgument(itemIdOrIndex)) {
+			return false
+		}
+
+		const index = Number.parseInt(itemIdOrIndex)
+
+		if (!Number.isNaN(index) && index > 0 && index <= list.length) {
+			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+			// @ts-ignore
+			const pk = list[index - 1][primaryKeyName]
+			if (typeof pk === 'string') {
+				return pk
+			} else {
+				throw Error(`invalid type ${typeof pk} for primary key`  +
+					` ${primaryKeyName} in ${JSON.stringify(list[index - 1])}`)
+			}
+		} else {
+			return false
+		}
+	}
+
+	const itemIdOrIndex: string = (await inquirer.prompt({
+		type: 'input',
+		name: 'itemIdOrIndex',
+		message: 'Enter id or index',
+		validate: (input) => {
+			return convertToId(input)
+				? true
+				: `Invalid id or index ${itemIdOrIndex}. Please enter an index or valid id.`
+		},
+	})).itemIdOrIndex
+	const inputId = convertToId(itemIdOrIndex)
+	if (inputId === false) {
+		throw Error(`unable to convert ${itemIdOrIndex} to id`)
+	}
+	return inputId
+}
+
+
+export type SelectingCommand<L> = SmartThingsCommandInterface & Sorting & CommonListOutputProducer<L>
+
+/**
+ * Process a command that selects and item (e.g. a device), performs some action on that device.
+ * This method is mainly here to combine shared code for the methods below but could be used
+ * elsewhere in circumstances where unusual output might be necessary.
+ *
+ * @param command The command itself which conforms to the `SelectingCommand` type.
+ * @param id The id of the item to act on, or undefined if the user should be asked to select one.
+ * @param listFunction A function that returns a list of all items which will be used to build the
+ *   list of the user to choose from when `id` is undefined.
+ * @param actionFunction The function that performs the action.
+ * @param getIdFromUser A function that queries the user into a for an item given a list of them.
+ *
+ * @returns a tuple containing the id of the item acted on and the result of the action function.
+ */
+export async function selectAndActOnGeneric<ID, O, L>(command: SelectingCommand<L>,
+		id: ID | undefined, listFunction: ListDataFunction<L>, actionFunction: ActionFunction<ID, O>,
+		getIdFromUser: IdRetrievalFunction<ID, L>): Promise<[ID, O]> {
+	let computedId: ID
+	if (id) {
+		computedId = id
+	} else {
+		const list = await outputList(command, listFunction, true)
+		if (list.length === 0) {
+			// Nothing was found; user was already notified.
+			command.exit(0)
+		}
+		computedId = await getIdFromUser(command.primaryKeyName, list)
+	}
+	const updatedItem = await actionFunction(computedId)
+	return [computedId, updatedItem]
```

Comment:
I'm generally not a fan of tuples (with their unnamed fields) but they seem kind of perfect for this use-case. Anyone else have opinions on this?

Reply:
It works and is type safe thanks to TS. I say 👍 .